File: js/257_BinaryTreePaths.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string[]}
 */
var binaryTreePaths = function(root) {
    var ret = [];
    if(root === null) return ret;
    pathTree(root,root.val.toString(),ret);
    return ret;
};

var pathTree = function(root,pre,ret){
    if(root.left === null && root.right === null){
        ret.push(pre);
        return;
    }
    if(root.left !== null) pathTree(root.left, pre+"->"+root.left.val.toString(), ret);
    if(root.right !== null) pathTree(root.right, pre+"->"+root.right.val.toString(), ret);
};

File: cpp/257_BinaryTreePaths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        if(!root) return ret;
        pathTree(root, std::to_string(root->val));
        return ret;
    }
private:
    vector<string> ret;
    
    void pathTree(TreeNode* root,string pre){
        if(!root->left && !root->right){
            ret.push_back(pre);
            return; 
        }
        if(root->left) pathTree(root->left, pre + "->" + std::to_string(root->left->val));
        if(root->right) pathTree(root->right,pre + "->" + std::to_string(root->right->val));
    }
};